### PR TITLE
basic server implementation and front end example

### PR DESCRIPTION
**Pull request by Jhaines1988**
This is a basic server setup that has one  post route that accepts all queries. 

On the front end , in the App component, I've created a formatter. If we are to use it, it is very important that the data be put in the formatter function in the correct order.  I've attached some screenshots  and will provide some description. 

If we decide to use this instead of creating many backend routes for our api calls, we must be very careful about distinguishing parameters that must go in an object and are attached to the body of the request,  versus parameters that can go directly in the query string. 

Why a formatter?  Because  it prevents us from having to write cumbersome query-strings for get requests, though it is still some work to format the request correctly.  The function is reusable and I have tested it with all the attached screenshots. Even the PUT review as helpful worked, though I had to verify it by querying the review again because the status is says it will send back is 201 , but it send 200.    At any rate, the review I marked as helpful did change so I know it works in at least one case.  POST requests have not been tested yet. If someone would like to take that on , let me know. 

As a rule, Whenever you see /:someRoute  - that means you can directly place it in the query string.   EDIT: In the case of our function, that parameter would be apiRoute.  EX  a Get request for  products/:product_id/styles  could be formatted  like this: 

format body ( 'GET',  '/products/47425/styles')   
I can leave out the other two parameters  **(params, data)** because they don't matter and Axios doesn't care if empty objects are attached to the request unless you're trying to do something specific with them. 

if something  like product_id is  listed as a parameter  such as  in GET /reviews ,  then you must create an object and put it into the appropriate spot for the query formatter. For get requests, you d put this in the formatter: 

formatBody('GET', '/reviews/ ,{product_id: 47425}) 

we would leave the last argument "data") out because for Get requests we attach parameters. 

For Post / PUT request, you would use this function to attach the relevant data as an object in the correct slot; 

EX  formatBody('POST',  /reviews/,  null, {product_id: 47425, rating: 3, body: 'somereviewtext', etc....} 

notice that in the space allotted for params in the function, we pass in NULL.   Hopefully this makes sense. 

I do not expect this to be merged into main as is.  Notice in our APP component i've rendered some data from component did mount and put it on the page.  I have kept it as an object for now   as that is whats returned. However, we really need to think about what data we need to get right away, and especially concerning is if we should use component did mount at all. 

The alternative would be to hardcode the product id and pass it down through props, along with a bunch of methods that set the state of particular variables.   

So to recap, we need a high-level strategy / overview of what we all need, and I was hoping we could have that discussion tomorrow if possible, though I am around and working today. 

PS: Sorry for the amount of detail, this is probably a relatively large PR, I'll work on smaller from here on out: 

Joe

Screen Shot 2021-09-10 at 11 47 16 AM

Screen Shot 2021-09-10 at 11 14 24 AM
Screen Shot 2021-09-10 at 11 11 33 AM
Screen Shot 2021-09-10 at 11 08 50 AM
Screen Shot 2021-09-10 at 11 08 13 AM
Screen Shot 2021-09-10 at 11 10 47 AM
Screen Shot 2021-09-10 at 11 20 01 AM

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,12 +1,40 @@
-const path = require("path")
-const express = require("express");
+const path = require('path');
+const express = require('express');
+const config = require('./config');
+const token = config.token;
+const apiUrl = 'https://app-hrsei-api.herokuapp.com/api/fec2/hr-rpp';
```
Comment (ashequl93) [inline server.js:5]: To follow the convention let's name all constant UPPERCASE. For example, APIURL. So we can identify it later quickly.

File: server.js [hunk 2]
```diff
@@ -1,12 +1,40 @@
...
+app.post('/api/*', (req, res)=>{
+  console.log('body', req.body);
+
+  let options = req.body;
+  options.headers.Authorization = token;
+
+
+  axios(options).then((results)=>{
+    console.log('API Results:', results.data);
+    res.send(results.data);
```
Comment (ashequl93) [inline server.js:26]: It would be better to send **statusCode** along with the data. Ex: **_res.status(statusCode).send(data)_**

File: server.js [hunk 3]
```diff
@@ -1,12 +1,40 @@
...
+  }).catch((err)=>{
+    console.log('API Error:', err);
+    res.send(err);
```
Comment (ashequl93) [inline server.js:29]: Same for error **statusCode**

File: client/src/components/App.jsx [hunk 4]
```diff
@@ -1,17 +1,61 @@
...
+
+  }
+
+  formatBody(method, apiRoute, params, data) {
+
+    let bodyObj = {
+      method: method,
+      url: `https://app-hrsei-api.herokuapp.com/api/fec2/hr-rpp${apiRoute}`,
+      data: data,
+      params: params,
```
Comment (ashequl93) [inline client/src/components/App.jsx:22]: If **data** and **params** are empty for some call, we can put some **default value** for them instead of **undefined**. Ex: **_data: data || {}_** or we can also put in the method signature. **_formatBody(method, apiRoute, params, data={})_**

File: server.js (hunk 1 above)
Comment (Jhaines1988) [reply server.js:5]: Thanks Ash , 

In the above, you are  only referring to  CONFIG, TOKEN, and API , correct? 

I've never uppercased express, but am open if thats what you are suggesting.

File: server.js (hunk 2 above)
Comment (Jhaines1988) [reply server.js:26]: Will do for  this and below comment.

File: client/src/components/App.jsx [hunk 5]
```diff
@@ -1,17 +1,61 @@
...
+      headers: {Authorization: ''}
+    };
+
+
```
Comment (ashequl93) [inline client/src/components/App.jsx:26]: There are some extra spaces all over the place. We can remove that by using any **_formatter_** or manually.

File: client/src/components/App.jsx (hunk 4 above)
Comment (Jhaines1988) [reply client/src/components/App.jsx:22]: Agreed.  I prefer to put them  as defaults in the actual function to minimize user error.  Sound Ok?

File: server.js [hunk 6]
```diff
@@ -1,12 +1,40 @@
-const path = require("path")
-const express = require("express");
+const path = require('path');
+const express = require('express');
+const config = require('./config');
+const token = config.token;
```
Comment (ashequl93) [inline server.js:4]: For constant, we can use UPPERCASE to follow the convention.

**Review by ashequl93 — CHANGES_REQUESTED** [review]

File: client/src/components/App.jsx (hunk 4 above)
Comment (ashequl93) [reply client/src/components/App.jsx:22]: Sounds good to me.

File: server.js (hunk 1 above)
Comment (ashequl93) [reply server.js:5]: We don't need to uppercase config. Just TOEKN & APIURL

**Review by aaronfife — APPROVED** [review]

**Review by ashequl93 — APPROVED** [review]
